### PR TITLE
ActiveRecord/Association Tests & Support

### PR DESCRIPTION
**Pull request by dlcarter**
*What this PR accomplishes:*

* Support for nested serializers & views for associations
  e.g. `association :vehicles, serializer: VehicleViewSerializer, view: :extended`
* Extended tests for associations and Active Record lookups
* Removes ActiveRecord as a foundational requirement for the serializer
* Removes our dependency on ActiveRecord for the base serializer tests, greatly speeding them up

### CODE REVIEW

File: lib/painted_rabbit/base.rb [hunk 1]
```diff
@@ -22,21 +24,24 @@ def self.field(method, options = {})
...
     def self.render(object, view: :default)
+      jsonify(hashify(object, view: view))
+    end
+
+    def self.hashify(object, view:)
       unless views.keys.include? view
         raise PaintedRabbitError, "View '#{view}' is not defined"
       end
-      if object.respond_to? :each
+      output_hash = if object.respond_to? :each
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: Unused assign?

File: lib/painted_rabbit/base.rb [hunk 2]
```diff
@@ -112,17 +113,23 @@ def self.include_in(view_list)
...
+        JSON.generate(blob)
+      end
+    end
+
     # I had to test this, so a note for later, class level instance variables
     # are not mutated at the parent class level when they are changed at the
     # inherited class
     def self.tracked_fields
       @tracked_fields ||= []
     end
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: This is not used anywhere else... looks safe to remove

File: lib/painted_rabbit/base.rb [hunk 3]
```diff
@@ -22,21 +24,24 @@ def self.field(method, options = {})
...
+      jsonify(hashify(object, view: view))
+    end
+
+    def self.hashify(object, view:)
       unless views.keys.include? view
         raise PaintedRabbitError, "View '#{view}' is not defined"
       end
-      if object.respond_to? :each
+      output_hash = if object.respond_to? :each
         if object.respond_to? :select # TODO: Change to more explicitely test for AR
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: How about `is_a?`

File: lib/painted_rabbit/base.rb [hunk 4]
```diff
@@ -22,21 +24,24 @@ def self.field(method, options = {})
...
           select_columns = (active_record_attributes(object) &
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: We should refactor this array/AR lookup logic to give it a name!

File: lib/painted_rabbit/base.rb [hunk 5]
```diff
@@ -22,21 +24,24 @@ def self.field(method, options = {})
...
+                                           AssociationSerializer.bleh,
                                            options.merge(association: true))
       end
     end
 
     def self.render(object, view: :default)
+      jsonify(hashify(object, view: view))
+    end
+
+    def self.hashify(object, view:)
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: Name ideas: `normalize` ?

File: lib/painted_rabbit/base.rb [hunk 6]
```diff
@@ -112,17 +113,23 @@ def self.include_in(view_list)
       current_views = [:default]
     end
 
+    private
+
+    def self.jsonify(blob)
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: Possibly name this `serialize` ?

File: lib/painted_rabbit/base.rb [hunk 7]
```diff
@@ -112,17 +113,23 @@ def self.include_in(view_list)
       current_views = [:default]
     end
 
+    private
```
Comment (dlcarter) [inline lib/painted_rabbit/base.rb]: these methods won't be private unless we use `private_class_method`

File: test/active_record_test.rb [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      end
+    end
+
+    user = create(:user)
+
+    expected_default_view = JSON.generate({
+      id: user.id,
+      email: user.email
+    })
+    assert_equal(expected_default_view, ViewUserSerializer.render(user))
```
Comment (dlcarter) [inline test/active_record_test.rb]: Consolidate setup and segregate from assertions

File: test/painted_rabbit_test.rb [hunk 9]
```diff
@@ -1,30 +1,32 @@
...
+
+  def test_truth
     assert_kind_of Module, PaintedRabbit
   end
 
-  test "it renders based on a templates class" do
+  def test_templates_class
     my_obj = OpenStruct.new(id: 1, name: 'Meg')
-    SimpleRabbit = Class.new(PaintedRabbit::Base) do
+    simple_rabbit = Class.new(PaintedRabbit::Base) do
```
Comment (dlcarter) [inline test/painted_rabbit_test.rb]: let's name it `simple_rabbit_class`

File: lib/painted_rabbit/base.rb (hunk 5 above)
Comment (philipqnguyen) [inline lib/painted_rabbit/base.rb]: Should `hashify` be a public method?

File: lib/painted_rabbit/serializer.rb [hunk 10]
```diff
@@ -10,7 +10,7 @@ def self.bleh
     @_blah
```
Comment (mcclayton) [inline lib/painted_rabbit/serializer.rb]: I might have missed something, but why the `bleh` and `@_blah`?
Placeholders?

Comment (dlcarter) [reply lib/painted_rabbit/serializer.rb]: Yes - that code wasn't introduced in this PR, it is original to the repo. We plan on refactoring/fixing this in the very next PR.

**Pull request merged**